[Blog] Testable IAM Policy documents - DO NOT MERGE

tdpauw opened the pull request:
Hello,

A time ago, I published a Typescript library [`@thinkinglabs/aws-iam-policy`](https://github.com/thinkinglabs/aws-iam-policy) on Pulumi's Slack channel. The library allows for the creation of IAM Policy documents programmatically using an object model. The idea behind this was to be able to unit test IAM Policies more easily.

@jaxxstorm seemed to like the library and asked if I wanted to contribute an article for Pulumi's blog.

So, here is my article. It explains why I've implemented this library.
Let me know if this is valuable for Pulumi.

Note: I did not have much inspiration for the meta image. I took a screenshot of an IAM Policy JSON document, but it shows a bit small on the blog article list. If you have a better idea, feel free to advise.

Cheers,
Thierry

spara commented:
> Hi Thierry,
> 
> Thank you for the contribution. I'll review it shortly.

I have a number of comments and editorial suggestions. Could you follow these instructions to let me make commits to your fork's branch? 

https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

tdpauw commented:
> I have a number of comments and editorial suggestions. Could you follow these instructions to let me make commits to your fork's branch?
>
> https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

Apparently, "allow edits from maintainers" is not available for a PR coming from an organisation repository. 🤷‍♂️ 
So, I granted you write access to the source repository. I hope this helps.

spara commented:
> > I have a number of comments and editorial suggestions. Could you follow these instructions to let me make commits to your fork's branch?
> > https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork
> 
> Apparently, "allow edits from maintainers" is not available for a PR coming from an organisation repository. 🤷‍♂️
> So, I granted you write access to the source repository. I hope this helps.

That reworked, please review the suggested changes.

tdpauw commented:
> please review the suggested changes.

Wow, that looks far better. Thank you, @spara ! 😊 

I took the liberty to fix some typos and missing spaces.

I also took the liberty to reintroduce a "However" and "But" at two places because, to me, it feels more natural. But then again, I am not a native English speaker. If you don't like it, please remove it.

When rereading I realised there was a sentence missing in front of the S3 bucket policy unit-test example. And some code was missing in the S3 bucket policy example.

Let me know what you think of it.

spara commented on themes/default/content/blog/testable-iam-policy-documents/index.md:
```diff
@@ -0,0 +1,322 @@
+    - IAM
+---
+
+I was relieved to find Pulumi. Finally, we have testable Infrastructure as Code. We can write fast unit tests that we can execute locally without needing the cloud. However, I was disappointed that Pulumi does not have a proper API for manipulating AWS IAM Policy documents.
```
```suggestion
I was relieved to find Pulumi. Finally, we have testable Infrastructure as Code. We can write fast unit tests that we can execute locally without needing the cloud. However, Pulumi doesn't currently have a full representation of IAM policy documents. Fortunately, it was relatively easy to build a library which did this!
```

spara commented on themes/default/content/blog/testable-iam-policy-documents/index.md:
````diff
@@ -0,0 +1,322 @@
+});
+```
+
+However, it is perfectly possible to pass an invalid IAM Policy document because there is no validation. You would only notice if it is invalid the minute the policy is applied in the AWS cloud. That is an unreasonably long feedback loop, incurring a significant amount of waiting and time to correction.
````
```suggestion
However, it is perfectly possible to pass an invalid IAM Policy document because there is no validation. You would only notice if it is invalid the minute the policy is applied in the AWS cloud. That creates a long feedback loop, incurring a significant amount of waiting and time to correction.
```

spara commented on themes/default/content/blog/testable-iam-policy-documents/index.md:
```diff
@@ -0,0 +1,322 @@
+common syntax errors, reduces the feedback cycle and increases
+your delivery throughput.
+
+Having to pass a JSON as a policy document was a bit disappointing.
```
```suggestion
Having to pass a JSON as a policy document didn't feel optimal.
```

spara commented on themes/default/content/blog/testable-iam-policy-documents/index.md:
```diff
@@ -0,0 +1,322 @@
+I work in the financial industry and compliance is important. So, I was in search of something that allowed me to easily unit test IAM Policy documents, preferably at the Statement level, which would help us to adhere to security requirements.
+
+Before reinventing the wheel, I searched for existing packages in
+JavaScrip for manipulating IAM Policy documents.
```
```suggestion
JavaScript for manipulating IAM Policy documents.
```

spara reviewed (changes requested):
@jaxxstorm suggest that you add a summary or conclusion to finish the post. For example:

## Conclusion

As you can see, you can extend a Pulumi program with existing Node modules or write your own. By testing IAM policies before deployment, we can avoid errors before deployments and provide guardrails for developers.

spara commented on themes/default/content/blog/testable-iam-policy-documents/index.md:
```diff
@@ -0,0 +1,322 @@
+date: 2021-05-06
+draft: true
+meta_desc:
+meta_image: meta.png
```
```suggestion
meta_image: testable_iam_policy.png
```

spara reviewed (changes requested):
added a new meta image
![testable_iam_policy](https://user-images.githubusercontent.com/638672/117463835-b3c2e380-af15-11eb-9e84-561daf5f87a7.png)

tdpauw commented:
> @jaxxstorm suggest that you add a summary or conclusion to finish the post.

Good point. Indeed, I also felt there was something missing at the end. I've added the suggested conclusion and improved it a little bit.

I've also included your suggested modifications and the meta image (which is indeed better).

Lastly, I've changed the publication date to May 10th and removed `draft: true` from the frontmatter.

Well, I think we have come to something. Isn't it? 😁  Thank you for your help!

spara reviewed (changes requested):
Minor grammatical and style edits. Otherwise looks great and ready for publishing.

tdpauw commented:
I've seen that. I'm ok with the edits. Makes total sense.

spara commented:
> > @jaxxstorm suggest that you add a summary or conclusion to finish the post.
> 
> Good point. Indeed, I also felt there was something missing at the end. I've added the suggested conclusion and improved it a little bit.
> 
> I've also included your suggested modifications and the meta image (which is indeed better).
> 
> Lastly, I've changed the publication date to May 10th and removed `draft: true` from the frontmatter.
> 
> Well, I think we have come to something. Isn't it? 😁 Thank you for your help!

Thierry,

It looks like there was a bad rebase over the weekend and added 60+ files to this PR. There are a couple options to fix this and ensure you get contributor credit.

1.  Copy the folder containing the blog post and the author entry into some other place on disk, delete the branch locally, fetch latest master and create a new branch by the same name, copy the blog folder and author file in, and commit and force-push the same-named branch to their repo.
2. I can close this PR and open a new PR with you as a co-contributor so you can get the github contributor credit.

Let me know which one you want to do.

sophia

tdpauw commented:
Oh, I've missed that message from you.
Damned it, I might have done something wrong when fetching the latest changes from the upstream repo.

Let's go for the second option. Force push feels awkward. I'm really sorry for the inconvenience.

spara commented:
closed, moved to #216
